### PR TITLE
feat: host health check api

### PR DESCRIPTION
**Pull request by adrysn**
This PR adds manager/agent host health check API at `/manager/health`.

Internal Ticket: OP#664
Related: https://github.com/lablup/backend.ai-agent/pull/241, https://github.com/lablup/backend.ai-common/pull/60, https://github.com/lablup/backend.ai-client-py/pull/145

### CODE REVIEW

File: src/ai/backend/gateway/manager.py [hunk 1]
```diff
@@ -223,6 +231,102 @@ async def perform_scheduler_ops(request: web.Request, params: Any) -> web.Respon
     return web.Response(status=204)
 
 
+@superadmin_required
+@check_api_params(
+    t.Dict({
+        tx.MultiKey('agent_ids', default=[]): t.List(t.Null | t.String),
+        t.Key('with_all_agents', default=False): t.ToBool,
```
[the remark this answers is not in the document]
Comment (achimnol) [reply src/ai/backend/gateway/manager.py:238]: To handle "string-or-list-of-strings" in a more consistent way, we could add a new trafaret as `ai.backend.common.validators.XList`, which accepts either an item or a list of items and always returns the list of items.

**Review by achimnol — CHANGES_REQUESTED** [review]

File: src/ai/backend/gateway/manager.py [hunk 2]
```diff
@@ -223,6 +231,102 @@ async def perform_scheduler_ops(request: web.Request, params: Any) -> web.Respon
...
+        return web.json_response(result)
+
+    # ## Get agent host information
+    async def _agent_health_check(sess: aiohttp.ClientSession, agent_id: str) -> dict:
+        none_result: dict = {agent_id: {}}
+        watcher_info = await get_watcher_info(request, agent_id)
+        if not watcher_info:
+            return none_result
+        watcher_url = watcher_info['addr'] / 'health'
+        with _timeout(10.0):
```
[the remark this answers is not in the document]
Comment (adrysn) [reply src/ai/backend/gateway/manager.py]: Replaced `_timeout` with `aiohttp.ClientTimeout` per request.

File: src/ai/backend/gateway/manager.py [hunk 3]
```diff
@@ -223,6 +231,102 @@ async def perform_scheduler_ops(request: web.Request, params: Any) -> web.Respon
...
+    connector = aiohttp.TCPConnector()
+    async with aiohttp.ClientSession(connector=connector) as sess:
+        scheduler = await aiojobs.create_scheduler(limit=4)
+        try:
+            jobs = await asyncio.gather(*[
+                scheduler.spawn(_agent_health_check(sess, aid)) for aid in params['agent_ids']
+            ])
+            agent_infos = await asyncio.gather(*[job.wait() for job in jobs])
+        finally:
+            await scheduler.close()
```
[the remark this answers is not in the document]
Comment (adrysn) [reply src/ai/backend/gateway/manager.py]: Replaced aiojobs with `aiotools.TaskGroup`.

File: src/ai/backend/gateway/manager.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (adrysn) [reply src/ai/backend/gateway/manager.py]: I replaced `MultiKey` with `Key` to accept either one of `str` or `list[str]`.

But for the second comment, isn't `tx.MultiKey` already do the job you described? With `tx.MultiKey` validator, it accepts both `str` and `list[str]` and always returns `list[str]`. For example, when I request `agents_ids` with `*`, it is converted to `['*']`. Please let me know if I mis-understood the comment!

[the remark this answers is not in the document]
Comment (achimnol) [reply src/ai/backend/gateway/manager.py]: Ah! I missed about `tx.MultiKey`!
Ok, then I'll accept this as-is.

File: src/ai/backend/gateway/manager.py [hunk 4]
```diff
@@ -291,35 +298,38 @@ async def health_check(request: web.Request, params: Any) -> web.Response:
...
-            ])
-            agent_infos = await asyncio.gather(*[job.wait() for job in jobs])
-        finally:
-            await scheduler.close()
+        tasks = []
+        async with TaskGroup() as tg:
+            sem = asyncio.Semaphore(10)
+            for aid in agent_ids:
+                async with sem:
+                    tasks.append(tg.create_task(_agent_health_check(sess, aid)))
```
Comment (achimnol) [inline src/ai/backend/gateway/manager.py:329]: The semaphore must be applied inside `_agent_health_check()`!

**Review by achimnol — CHANGES_REQUESTED** [review]

File: src/ai/backend/gateway/manager.py (hunk 4 above)
Comment (adrysn) [reply src/ai/backend/gateway/manager.py]: Thanks. Moved the semaphore inside `_agent_health_check()`

**Comment by adrysn** [comment]
This is deprecated in favor of https://github.com/lablup/backend.ai-watcher.

**Pull request closed without merging**